cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -435,27 +444,33 @@ public ProgramRunClusterStatus recordProgramDeprovisioning(ProgramRunId programR
     }
 
     ProgramRunClusterStatus clusterStatus = existing.getCluster().getStatus();
-    if (clusterStatus != ProgramRunClusterStatus.PROVISIONED) {
+    if (clusterStatus != ProgramRunClusterStatus.PROVISIONED && clusterStatus != ProgramRunClusterStatus.PROVISIONING) {
       LOG.warn("Ignoring unexpected request to transition program run {} from cluster state {} to cluster state {}.",
                programRunId, clusterStatus, ProgramRunClusterStatus.DEPROVISIONING);
       return null;
     }
-    if (!existing.getStatus().isEndState()) {
+    if (clusterStatus == ProgramRunClusterStatus.PROVISIONED && !existing.getStatus().isEndState()) {
       LOG.warn("Ignoring unexpected request to transition program run {} from program state {} to cluster state {}.",
                programRunId, existing.getStatus(), ProgramRunClusterStatus.DEPROVISIONING);
       return null;
     }
 
+    delete(existing);
```

Comment:
What does this do? Deletes the run record? I thought we never delete run record.

Reply:
it's because the run record keys are weird. They use different row keys depending on if its starting, running, or completed.